### PR TITLE
fixed connection leak issue

### PR DESCRIPTION
**Pull request by yongliu-mdsol**
**Issue:**
On enqueuing a job with BackgroudJobClient, HangfireTags initiates another connection without disposing properly. It exhausted all DB Connections. In our test case, we have 100 max connections. It failed when we enqueue 100+ jobs. 

image

### CODE REVIEW

**Comment by yongliu-mdsol** [comment]
@erwin-faceit @face-it
Thank you for merging my previous PR.

We did another round of testing and found an issue in the HangfireTag. It's actually critical enough that it breaks other hangfire activities. Please take a look at this PR.

**Comment by yongliu-mdsol** [comment]
@erwin-faceit @face-it
FYI
Another upcoming change we've done is changing the tag labels into searchable dropdown list.
It saves a lot of  real estate on the UI page. 

image
image

File: src/Hangfire.Tags/States/CreateJobFilter.cs [hunk 1]
```diff
@@ -13,7 +13,15 @@ public void OnCreating(CreatingContext filterContext)
 
         public void OnCreated(CreatedContext filterContext)
         {
-            var mi = filterContext.BackgroundJob.Job.Method;
+            // BackgroundJob can be null from CreatedContext
+            var mi = filterContext?.BackgroundJob?.Job?.Method;
```
Comment (ryan-dayrit) [inline src/Hangfire.Tags/States/CreateJobFilter.cs]: No exception NullReferenceException throw or log statement?

File: src/Hangfire.Tags/Storage/TagsStorage.cs [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 namespace Hangfire.Tags.Storage
 {
-    internal class TagsStorage : ITagsStorage, ITagsMonitoringApi
+    internal class TagsStorage : ITagsStorage, ITagsMonitoringApi, IDisposable
```
Comment (rdayrit-mdsol) [inline src/Hangfire.Tags/Storage/TagsStorage.cs:10]: public?

File: src/Hangfire.Tags.PostgreSql/PostgreSqlTagsServiceStorage.cs [hunk 3]
```diff
@@ -56,7 +57,7 @@ public IEnumerable<string> SearchTags(string tag, string setKey)
             return monitoringApi.UseConnection(connection =>
             {
                 var sql =
-                    $@"select Value from {_options.SchemaName}.Set s where s.Key like @setKey + ':%' + @tag + '%'";
+                    $@"select Value from {_options.SchemaName}.Set s where s.Key like (@setKey || ':%' || @tag || '%')";
```
Comment (danielloganking) [inline src/Hangfire.Tags.PostgreSql/PostgreSqlTagsServiceStorage.cs:60]: Should this be a separate commit since it fixes a different issue than the connection exhaustion? (Can still be this PR)

File: src/Hangfire.Tags/States/CreateJobFilter.cs (hunk 1 above)
Comment (yongliu-mdsol) [reply src/Hangfire.Tags/States/CreateJobFilter.cs]: fixed.

File: src/Hangfire.Tags.PostgreSql/PostgreSqlTagsServiceStorage.cs (hunk 3 above)
Comment (yongliu-mdsol) [reply src/Hangfire.Tags.PostgreSql/PostgreSqlTagsServiceStorage.cs:60]: yup

**Review by ryan-dayrit — APPROVED** [review]
LGTM

**Review by danielloganking — APPROVED** [review]
LGTM

**Review by erwin-faceit — COMMENTED** [review]
Great change, and an excellent find! Thank you very much.

I see that you've actually "mixed" multiple changes into one PR, and I think we should clean that up.

1. You've added a check in the CreateJobFilter, which throws an ArgumentException instead of an NRE.
2. You've fixed the disposal
3. You've changed some interfaces to public, I think for the upcoming change.

I'll add some comments to the individual changes, but to keep it clean, I think it's best to separate these into individual PR's.

Let me know what you think.

File: src/Hangfire.Tags/States/TagsCleanupStateFilter.cs [hunk 4]
```diff
@@ -8,15 +8,17 @@ internal class TagsCleanupStateFilter : IApplyStateFilter
     {
         public void OnStateApplied(ApplyStateContext context, IWriteOnlyTransaction transaction)
         {
-            var expireTransaction = new TagExpirationTransaction(context.Storage, (JobStorageTransaction) transaction);
-            var jobid = context.BackgroundJob.Id;
+            using (var expireTransaction = new TagExpirationTransaction(context.Storage, (JobStorageTransaction)transaction))
```
Comment (erwin-faceit) [inline src/Hangfire.Tags/States/TagsCleanupStateFilter.cs:11]: LGTM

File: src/Hangfire.Tags/Storage/ITagsMonitoringApi.cs [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 namespace Hangfire.Tags.Storage
 {
-    internal interface ITagsMonitoringApi
+    public interface ITagsMonitoringApi
```
Comment (erwin-faceit) [inline src/Hangfire.Tags/Storage/ITagsMonitoringApi.cs:7]: Not required for the connection leak issue

File: src/Hangfire.Tags/Storage/ITagsStorage.cs [hunk 6]
```diff
@@ -7,7 +7,7 @@ namespace Hangfire.Tags.Storage
     /// <summary>
     /// Abstraction over Hangfire's storage API
     /// </summary>
-    internal interface ITagsStorage : IDisposable
+    public interface ITagsStorage : IDisposable
```
Comment (erwin-faceit) [inline src/Hangfire.Tags/Storage/ITagsStorage.cs:10]: Not required for the connection leak issue

File: src/Hangfire.Tags/Storage/TagExpirationTransaction.cs [hunk 7]
```diff
@@ -22,6 +22,7 @@ public TagExpirationTransaction(TagsStorage tagsStorage, JobStorageTransaction t
         public void Dispose()
         {
             _transaction.Dispose();
+            _tagsStorage.Dispose();
```
Comment (erwin-faceit) [inline src/Hangfire.Tags/Storage/TagExpirationTransaction.cs:25]: LGTM

File: src/Hangfire.Tags/Storage/TagsStorage.cs (hunk 2 above)
Comment (erwin-faceit) [inline src/Hangfire.Tags/Storage/TagsStorage.cs:10]: Not required, since ITagsStorage already inherits from IDisposable

**Comment by erwin-faceit** [comment]
Actually, I'm fine with merging it as is, just let me know if my review is correct, so that I know why you changed internal to public, and handled the NRE. 

Thanks again for your great work on this Hangfire extension. Much appreciated!

File: src/Hangfire.Tags/Storage/ITagsMonitoringApi.cs (hunk 5 above)
Comment (yongliu-mdsol) [reply src/Hangfire.Tags/Storage/ITagsMonitoringApi.cs:7]: yup, absolutely, it should be a different PR.

File: src/Hangfire.Tags/Storage/ITagsStorage.cs (hunk 6 above)
Comment (yongliu-mdsol) [reply src/Hangfire.Tags/Storage/ITagsStorage.cs:10]: yup, absolutely, it should be a different PR.

**Comment by yongliu-mdsol** [comment]
@erwin-faceit 
Your comments are valid. I will definitely do that next time. 
I wasn't sure about your preference over multiple PRs. 
What's your package versioning strategy? Do we bump version on every PR? 
Please let me know because I have another upcoming PR to improve the UX. 

I changed the interfaces to public so that we can extend the Tags UI. (screenshots in this PR).

**Comment by erwin-faceit** [comment]
Alright, let's merge this change then. You're version strategy is fine. I'm used to using gitflow, and let gitversion handle the versionnumbers. I didn't implement it correctly in this project though, so I'm usually just updating the versionnumbers by hand...

Let's just bump the version on every change, it's fine by me.
